ASoC: SOF: IPC4: Add initial firmware handling support

ujfalusi opened the pull request:
Hi,

The series implements basic support for parsing the main firmware manifest for IPC4 based image.
Platform code must set up the `struct sof_ipc4_data` and sdev->private should be set to point to it since some parameters for parsing the firmware image is platform dependent.

plbossart commented on sound/soc/sof/ipc4-priv.h:
```diff
@@ -0,0 +1,26 @@
+/* SPDX-License-Identifier: (GPL-2.0-only OR BSD-3-Clause) */
```
commit message typo: "Add a struct'

plbossart commented on sound/soc/sof/ipc4-priv.h:
```diff
@@ -0,0 +1,26 @@
+ * @fw_modules: Array of base FW modules
+ * @manifest_fw_hdr_offset: FW header offset in the manifest
+ */
+struct sof_ipc4_data {
```
the name doesn't match the kernel-doc, and it's as vague as can be. 'data' could refer to just about anything.

sof_ipc4_fw_data?

ujfalusi replied:
Oops, I have overlooked this.

Yes, `sof_ipc4_fw_data` is more appropriate

plbossart commented on sound/soc/sof/ipc4-priv.h:
```diff
@@ -0,0 +1,26 @@
+struct sof_ipc4_data {
+	int num_fw_modules;
+	void *fw_modules;
+	u32 manifest_fw_hdr_offset;
```
add this first?

ujfalusi replied:
Sure.

ujfalusi commented:
Changes since v1:
- Typo fixed in commit messages (There were others than what @plbossart pointed out)
- Clarified some of the commit messages at the same time.
- sof_ipc4_data is renamed to sof_ipc4_fw_data and struct is re-arranged
- New patch is added to set up the sof_ipc4_fw_data for IPC4 capable platforms in case they are booted in IPC4 mode.

ranj063 commented on include/sound/sof/ext_manifest4.h:
```diff
@@ -0,0 +1,119 @@
+/* SPDX-License-Identifier: ((GPL-2.0 WITH Linux-syscall-note) OR BSD-3-Clause) */
```
@ujfalusi missing other people's co-developed. Originally written by Bard and Rander

ranj063 commented on sound/soc/sof/ipc4-loader.c:
```diff
@@ -0,0 +1,210 @@
+// SPDX-License-Identifier: (GPL-2.0-only OR BSD-3-Clause)
```
this one too

ujfalusi commented:
Changes since v2:
- Co-developed-by tags added

plbossart commented:
let's leave a chance to @ranj063 to review/merge.

ranj063 commented:
@ujfalusi there are a couple of patches in this series without your sign-off. Since it is your PR, we wont be gettingyour reviewed-by tag either. So maybe we'll have to fix it when we checkerry-pick to add it later? I dont want to block merging for it

ujfalusi commented:
> @ujfalusi there are a couple of patches in this series without your sign-off. Since it is your PR, we wont be gettingyour reviewed-by tag either. So maybe we'll have to fix it when we checkerry-pick to add it later? I dont want to block merging for it

Oops, well, if I will be sending it upstream they will get my SoB anyways. Sorry for missing that.
